src/sentry/services/hybrid_cloud/__init__.py

```diff
@@ -0,0 +1,322 @@
+import contextlib
+import logging
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from enum import Enum
+from typing import Dict, Generic, List, Mapping, Optional, Type, TypeVar, cast
+
+from sentry.models import Organization, OrganizationMember, OrganizationStatus
+
+logger = logging.getLogger(__name__)
+
+from django.db.models import F
+
+from sentry.silo import SiloMode
+
+
+class ProjectKeyRole(Enum):
+    store = "store"
+    api = "api"
+
+    def as_orm_role(self):
+        from sentry.models import ProjectKey
+
+        if self == ProjectKeyRole.store:
+            return ProjectKey.roles.store
+        elif self == ProjectKeyRole.api:
+            return ProjectKey.roles.api
+        else:
+            raise ValueError("Unexpected project key role enum")
+
+
+@dataclass
+class ApiProjectKey:
+    dsn_public: str = ""
+
+
+@dataclass
+class ApiOrganization:
+    slug: str = ""
+    id: int = -1
+
+
+class InterfaceWithLifecycle(ABC):
+    @abstractmethod
+    def close(self):
+        pass
+
+
+ServiceInterface = TypeVar("ServiceInterface", bound=InterfaceWithLifecycle)
+
+
+class DelegatedBySiloMode(Generic[ServiceInterface]):
+    """
+    Using a mapping of silo modes to backing type classes that match the same ServiceInterface,
+    delegate method calls to a singleton that is managed based on the current SiloMode.get_current_mode().
+    This delegator is dynamic -- it knows to swap the backing implementation even when silo mode is overwritten
+    during run time, or even via the stubbing methods in this module.
+
+    It also supports lifecycle management by invoking close() on the backing implementation anytime either this
+    service is closed, or when the backing service implementation changes.
+    """
+
+    _constructors: Mapping[SiloMode, Type[ServiceInterface]]
+    _singleton: Dict[SiloMode, ServiceInterface]
+
+    def __init__(self, mapping: Mapping[SiloMode, Type[ServiceInterface]]):
+        self._constructors = mapping
+        self._singleton = {}
+
+    @contextlib.contextmanager
+    def with_replacement(self, service: Optional[ServiceInterface], silo_mode: SiloMode):
+        prev = self._singleton
+        self.close()
+
+        if service:
+            self._singleton[silo_mode] = service
+            yield
+        else:
+            yield
+        self.close()
+        self._singleton = prev
+
+    def __getattr__(self, item: str):
+        cur_mode = SiloMode.get_current_mode()
+        if impl := self._singleton.get(cur_mode, None):
+            return getattr(impl, item)
+        if Con := self._constructors.get(cur_mode, None):
+            self.close()
+            return getattr(self._singleton.setdefault(cur_mode, Con()), item)
+
+        raise KeyError(f"No implementation found for {cur_mode}.")
+
+    def close(self):
+        for impl in self._singleton.values():
+            impl.close()
+        self._singleton = {}
+
+
+class ProjectKeyService(InterfaceWithLifecycle):
```

Comment:
Once the whole interface system is locked in, I think it would make sense to move each of these root service types and its "family" of subclasses to its own module. E.g.:
* `src/sentry/services/hybrid_cloud/projectkey.py`
* `src/sentry/services/hybrid_cloud/organization.py`
* etc.

Reply:
I absolutely agree.